enhancements/installer/agent-based-installation-in-hive.md

```diff
@@ -0,0 +1,647 @@
+clusters.
+
+Thus when a Cluster Creator creates a namespace and a new ClusterDeployment,
+that ClusterDeployment will need to consume Agents from a different namespace.
+But the Infra Owner may not want them to have the ability to consume any Agent
+from any namespace.
+
+To designate that a ClusterDeployment is authorized to consume Agents from a
+particular InstallEnv, a reference to that ClusterDeployment must be present
+in an array on the InstallEnv. The reference proves that someone with write
+access to the InstallEnv approves of associating it with the cluster.
+
+The restriction will be enforced by the agent-related controllers at the time
+of matching Agents to a particular ClusterDeployment. The controllers will
+still have access to all namespaces.
+
+InstallEnv and Agent resources are both namespaced. Agents will be created in
+the same namespace as the InstallEnv from which they were created.
+
+In the 4.8 timeframe, because "late binding" does not exist and thus each
+InstallEnv must be associated to a single cluster, this array of references
+will be required to have exactly one member. Once late binding is supported,
+that restriction will be removed. ("late binding" refers to the ability to
+create a discovery ISO and boot Agents with it prior to associating them with a
+cluster.)
+
+#### Create Cluster
+
+This scenario takes place on a hub cluster where hive and possibly RHACM are
+present. This scenario does not include centralized machine management.
+
+This scenario is an end-to-end flow that enables the user to specify everything
+up-front and then have an automated process create a cluster. Because the
+user's primary frame of reference is hardware-oriented, this flow enables them
+to specify Node attributes such as "role" with their BareMetalHost definition.
+Specifying those attributes together is particularly useful in a gitops
+approach.
+
+1. Infra Owner creates an InstallEnv resource. It can include fields such as egress proxy, NTP server, SSH public key, ... In particular it includes an Agent label selector, and a separate field of labels that will be automatically applied to Agents.
+1. Infra Owner creates BareMetalHost resources with corresponding BMC credentials. They must be labeled so that they match the selctor on the InstallEnv and must be in the same namespace as the InstallEnv.
+1. A new controller, the Baremetal Agent Controller, sees the matching BareMetalHosts and boots them using the discovery ISO URL found in the InstallEnv's status.
+1. The Agent starts up on each host and reports back to the assisted service, which creates an Agent resource in the same namespace as the InstallEnv. The Agent is automatically labeled with the labels that were specified in the InstallEnv's Spec.
+1. The Baremetal Agent Controller sets the Agent's Role field in its spec to "master" or "worker" if a corresponding label was present on its BareMetalHost.
```

Comment:
in RAN we can have other type of roles: worker-cnf, worker-cu, worker-du... and we know the profile that a node is going to take since the start. So, it is important to be able to specify the kind of worker that will be, so it picks the right ignition from start, and avoid not needed reconciliation/reboots. Currently our workflow is the following:
1. We deploy a control plane, enroll this cluster into ACM
2. From ACM, we apply all the cluster workloads (performance profile, create machine config pools, etc...)  for the cluster
3. ACM reconciles the changes and generates the right machine config pool into this cluster. As a consequence, we have a new ignition endpoint. It is not just config/master and config/worker, but we can have config/worker-cu or config/worker-du
4. We also deploy some autolabeler procedure, where we are capable of labeling the nodes with the right role before node is ready
5. When our worker node boots, we are pointing it to the worker-cu / worker-du ignition, so it picks all the right machine config items from scratch
6. Before node is ready, it gets labeled properly with the autolabeler. Then node joins as ready with worker, worker-cu labels. It is important to be able to support this workflow, and assisted installer is currently allowing to do it.

Reply:
Thanks for elaborating on that. For this proposal, we considered that to fall in the bucket of "remote worker node" functionality, which we listed as a non-goal because there were still some open questions about how to solve especially your step number 4. But we are tracking this use case outside this proposal, and the current approach being worked on is to use an admission webhook to [apply Node labels in-flight as they get created](https://issues.redhat.com/browse/KNIDEPLOY-4013).

Reply:
Ok, so we're saying this is already captured by:

> Remote Worker Node support requires solving specific problems that will be addressed in a separate proposal. For example, it requires the ability to ensure that the Node joins with an appropriate role such that it will not be schedulable for specific workloads, such as the ingress controller.

I guess the main point here though is that the user will supply the appropriate role via the AgentSpec.Role field in the case of boot-it-yourself, but in this case of using the bare metal agent controller it would come from a label on BareMetalHost?

But the "corresponding label [..] on its BareMetalHost" will have to be a well-known label we use for this purpose, which is quite different from the bare metal host selector on a Machine which can be based on any label

Reply:
The idea is to optionally express on a BareMetalHost what role it should have. If the BareMetalHost has a role defined, that just gets propagated to the Agent's Spec.

A user doing gitops for example wants to define their whole cluster layout up-front, and the BMH is the place they can define items such as role. It could be a label or an annotation, but since role seems like a useful grouping when doing a query, label seems like a good fit.

Reply:
Ok, have we defined anywhere what this label will be?

Reply:
@mgoldboi recently asked us to wait to decide on the exact label names. There are placeholders in code right now for various labels and annotations. I suspect this will be something like `agentinstall.openshift.io/role`. Ideas welcome!

Reply:
That's fine. I'd find this bullet more clear if it made it clear "the corresponding label" was something well-defined, not user-controlled. If you just said "for example, agentinstall.openshift.io/role" in parenthesis ...

Also, would you mind mentioning somewhere in here that the new API groups and labels will use a namespace like agentinstall.openshift.io but the exact name may change? Thanks

Reply:
Done. I'm adding the mention of the group name above in the "#### Kubernetes-native APIs for Agent Based Cluster Provisioning" section, just above describing InstallEnv, at line 202.